wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/kotlin/SourceRoots.kt

```diff
@@ -0,0 +1,165 @@
+  // TODO: Look at KotlinPlatformType when we get around to module dependencies and compatibility.
+  // We'll probably want to include that in the source so we can tell which source to rely on
+  // during dependency resolution.
+
+  return targets
+      .flatMap { target ->
+        if (target is KotlinAndroidTarget) {
+          val extension = project.extensions.getByType(BaseExtension::class.java)
+          return@flatMap extension.sourceRoots(project)
+              .map { source ->
+                val compilation = target.compilations.single { it.name == source.name }
+                return@map source.copy(
+                    name = "${target.name}${source.name.capitalize()}",
+                    sourceSets = source.sourceSets.map { "${target.name}${it.capitalize()}" } + "commonMain",
+                    registerTaskDependency = { task ->
+                      compilation.compileKotlinTask.dependsOn(task)
+                    }
+                )
+              }
+        }
+        return@flatMap target.compilations.mapNotNull { compilation ->
+          if (compilation.name.endsWith(suffix = "Test", ignoreCase = true)) {
+            return@mapNotNull null
+          }
+          Source(
+              type = target.platformType,
+              name = "${target.name}${compilation.name.capitalize()}",
+              variantName = (compilation as? KotlinJvmAndroidCompilation)?.name,
+              sourceDirectorySet = compilation.defaultSourceSet.kotlin,
+              sourceSets = compilation.allKotlinSourceSets.map { it.name },
+              registerTaskDependency = { task ->
+                (target as? KotlinNativeTarget)?.binaries?.forEach {
+                  it.linkTask.dependsOn(task)
+                }
+                compilation.compileKotlinTask.dependsOn(task)
+              }
+          )
+        }
+      }
+}
+
+private fun BaseExtension.sourceRoots(project: Project): List<Source> {
+  val variants: DomainObjectSet<out BaseVariant> = when (this) {
+    is AppExtension -> applicationVariants
+    is LibraryExtension -> libraryVariants
+    else -> throw IllegalStateException("Unknown Android plugin $this")
+  }
+  val sourceSets: Map<String, SourceDirectorySet?> = sourceSets
+      .associate { sourceSet ->
+        sourceSet.name to sourceSet.kotlin
+      }
+
+  return variants.map { variant ->
+    Source(
+        type = KotlinPlatformType.androidJvm,
+        name = variant.name,
+        variantName = variant.name,
+        sourceDirectorySet = sourceSets[variant.name]
+            ?: throw IllegalStateException("Couldn't find ${variant.name} in $sourceSets"),
+        sourceSets = variant.sourceSets.map { it.name },
+        registerTaskDependency = { task ->
+          // TODO: Lazy task configuration!!!
+          variant.registerJavaGeneratingTask(task.get(), task.get().outputDirectories)
+          project.tasks.named("compile${variant.name.capitalize()}Kotlin").dependsOn(task)
```

Comment:
is this coupling compilation to the generateProtos task?

if so, i feel pretty strongly on having this be the default.  our build speeds are terrible and re-generating protos in their modules constantly as part of the compile phase (gw app:assembleDebug) is a huge part of this.

Reply:
Our task should not run twice in a row. If there are cases when we run while it doesn't feel necessary we should fix. I think that's nice to have it as a task dependency.

Reply:
I think the brave strategy is too bind our task to the compilation and then to fix anything that sucks in ours.

Reply:
> Since generateProtos is basically a compilation task

No, it isn't, it's a code generation task.  We combine its output with other source files to proceed to compilation phase.  Perhaps I'm missing semantics.

> I think the brave strategy is too bind our task to the compilation and then to fix anything that sucks in ours.

We say that now, but our team is larger now and still growing (which is awesome!) so this affects more people.